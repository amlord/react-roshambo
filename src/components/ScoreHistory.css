.ScoreHistory__title {
    margin: 0;
    font-size: 1em;
}

.ScoreHistory__noGames {
    font-size: 0.9em;
    margin-top: 10px;
    color: #ccc;
}

.ScoreHistory__list {
    font-size: 0.9em;
}