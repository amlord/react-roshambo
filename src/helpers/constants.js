export const ROCK = 1;
export const PAPER = 2;
export const SCISSORS = 3;

export const WIN = 'win';
export const DRAW = 'draw';
export const LOSS = 'loss';

function _translateShape(shape) {
    switch(shape) {
        case ROCK:
            return 'Rock';
        case PAPER:
            return 'Paper';
        case SCISSORS:
            return 'Scissors';
    }
}

function _translateOutcome(outcome) {
    switch(outcome) {
        case WIN:
            return 'Win';
        case LOSS:
            return 'Lose';
        case DRAW:
            return 'Draw';
    }
}

export const Translate = {
    choice: function(shape) {
        return _translateShape(shape);
    },
    outcome: function(outcome) {
        return _translateOutcome(outcome);
    }
};