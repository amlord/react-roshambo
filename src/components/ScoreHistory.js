import React, { Component } from 'react';

import './ScoreHistory.css';
import { Translate } from '../helpers/Constants';

class ScoreHistory extends Component {

  renderHistory = () => {
    let { history } = this.props;

    return (
        <table className="gameHistory">
          {history.reverse().map((game, index) => {
              return (
                  <tr>
                      <td>{Translate.choice(game.player)}</td>
                      <td>{Translate.outcome(game.outcome)}</td>
                      <td>{Translate.choice(game.computer)}</td>
                  </tr>
              );
          })}
        </table>
    );
  }

  renderEmptyHistory = () => {
    return (
      <div>(no history)</div>
    );
  }

  render() {
    let { history } = this.props;

    return (
      <section className="ScoreHistory">
        {(history) ? this.renderHistory() : this.renderEmptyHistory()}
      </section>
    );
  }
}

export default ScoreHistory;
