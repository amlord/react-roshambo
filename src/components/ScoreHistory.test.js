import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';

import ScoreHistory from './ScoreHistory';
import { ROCK, PAPER, SCISSORS, WIN, DRAW, LOSS, Translate } from '../helpers/Constants';

describe('ScoreHistory', () => {
  const props = { history: [
    {player: ROCK, computer: PAPER, outcome: LOSS},
    {player: SCISSORS, computer: PAPER, outcome: WIN},
    {player: PAPER, computer: PAPER, outcome: DRAW}
  ]};

  describe('no history', () => {
    const scoreboard = shallow(<ScoreHistory history={[]} />);

    it('renders correctly', () => {
      expect(scoreboard).toMatchSnapshot();
    });

    it('does not show table rows', () => {
      expect( scoreboard.find('.ScoreHistory__list tr').length ).toEqual(0);
    });
  });

  describe('history', () => {
    const scoreboard = shallow(<ScoreHistory {...props} />);

    it('renders correctly', () => {
      expect(scoreboard).toMatchSnapshot();
    });

    it('shows 3 table rows', () => {
      expect( scoreboard.find('.ScoreHistory__list tr').length ).toEqual(3);
    });

    it('cell 1 shows player choice', () => {
      const txtRock = Translate.choice(ROCK);
      const txtPlayerChoice = scoreboard.find('.ScoreHistory__list tr').at(2).find('td').at(1).text();

      expect( txtPlayerChoice ).toEqual( txtRock );
    });

    it('cell 2 shows game outcome', () => {
      const txtLoss = Translate.outcome(LOSS);
      const txtOutcome = scoreboard.find('.ScoreHistory__list tr').at(2).find('td').at(2).text();
      
      expect( txtOutcome ).toEqual( txtLoss );
    });

    it('cell 3 shows computer choice', () => {
      const txtPaper = Translate.choice(PAPER);
      const txtComputerChoice = scoreboard.find('.ScoreHistory__list tr').at(2).find('td').at(3).text();
      
      expect( txtComputerChoice ).toEqual( txtPaper );
    });
  });
});