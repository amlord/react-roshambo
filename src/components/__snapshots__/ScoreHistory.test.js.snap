// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScoreHistory history renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ScoreHistory
    history={
        Array [
            Object {
              "computer": 2,
              "outcome": "loss",
              "player": 1,
            },
            Object {
              "computer": 2,
              "outcome": "win",
              "player": 3,
            },
            Object {
              "computer": 2,
              "outcome": "draw",
              "player": 2,
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <table
        className="gameHistory"
>
        <tr>
                <td>
                        Rock
                </td>
                <td>
                        Lose
                </td>
                <td>
                        Paper
                </td>
        </tr>
        <tr>
                <td>
                        Scissors
                </td>
                <td>
                        Win
                </td>
                <td>
                        Paper
                </td>
        </tr>
        <tr>
                <td>
                        Paper
                </td>
                <td>
                        Draw
                </td>
                <td>
                        Paper
                </td>
        </tr>
</table>,
      "className": "ScoreHistory",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <tr>
            <td>
                        Rock
            </td>
            <td>
                        Lose
            </td>
            <td>
                        Paper
            </td>
</tr>,
          <tr>
            <td>
                        Scissors
            </td>
            <td>
                        Win
            </td>
            <td>
                        Paper
            </td>
</tr>,
          <tr>
            <td>
                        Paper
            </td>
            <td>
                        Draw
            </td>
            <td>
                        Paper
            </td>
</tr>,
        ],
        "className": "gameHistory",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <td>
                Rock
</td>,
              <td>
                Lose
</td>,
              <td>
                Paper
</td>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Rock",
              },
              "ref": null,
              "rendered": "Rock",
              "type": "td",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Lose",
              },
              "ref": null,
              "rendered": "Lose",
              "type": "td",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Paper",
              },
              "ref": null,
              "rendered": "Paper",
              "type": "td",
            },
          ],
          "type": "tr",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <td>
                Scissors
</td>,
              <td>
                Win
</td>,
              <td>
                Paper
</td>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Scissors",
              },
              "ref": null,
              "rendered": "Scissors",
              "type": "td",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Win",
              },
              "ref": null,
              "rendered": "Win",
              "type": "td",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Paper",
              },
              "ref": null,
              "rendered": "Paper",
              "type": "td",
            },
          ],
          "type": "tr",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <td>
                Paper
</td>,
              <td>
                Draw
</td>,
              <td>
                Paper
</td>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Paper",
              },
              "ref": null,
              "rendered": "Paper",
              "type": "td",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Draw",
              },
              "ref": null,
              "rendered": "Draw",
              "type": "td",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Paper",
              },
              "ref": null,
              "rendered": "Paper",
              "type": "td",
            },
          ],
          "type": "tr",
        },
      ],
      "type": "table",
    },
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <table
          className="gameHistory"
>
          <tr>
                    <td>
                              Rock
                    </td>
                    <td>
                              Lose
                    </td>
                    <td>
                              Paper
                    </td>
          </tr>
          <tr>
                    <td>
                              Scissors
                    </td>
                    <td>
                              Win
                    </td>
                    <td>
                              Paper
                    </td>
          </tr>
          <tr>
                    <td>
                              Paper
                    </td>
                    <td>
                              Draw
                    </td>
                    <td>
                              Paper
                    </td>
          </tr>
</table>,
        "className": "ScoreHistory",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <tr>
              <td>
                            Rock
              </td>
              <td>
                            Lose
              </td>
              <td>
                            Paper
              </td>
</tr>,
            <tr>
              <td>
                            Scissors
              </td>
              <td>
                            Win
              </td>
              <td>
                            Paper
              </td>
</tr>,
            <tr>
              <td>
                            Paper
              </td>
              <td>
                            Draw
              </td>
              <td>
                            Paper
              </td>
</tr>,
          ],
          "className": "gameHistory",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <td>
                  Rock
</td>,
                <td>
                  Lose
</td>,
                <td>
                  Paper
</td>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Rock",
                },
                "ref": null,
                "rendered": "Rock",
                "type": "td",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Lose",
                },
                "ref": null,
                "rendered": "Lose",
                "type": "td",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Paper",
                },
                "ref": null,
                "rendered": "Paper",
                "type": "td",
              },
            ],
            "type": "tr",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <td>
                  Scissors
</td>,
                <td>
                  Win
</td>,
                <td>
                  Paper
</td>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Scissors",
                },
                "ref": null,
                "rendered": "Scissors",
                "type": "td",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Win",
                },
                "ref": null,
                "rendered": "Win",
                "type": "td",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Paper",
                },
                "ref": null,
                "rendered": "Paper",
                "type": "td",
              },
            ],
            "type": "tr",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <td>
                  Paper
</td>,
                <td>
                  Draw
</td>,
                <td>
                  Paper
</td>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Paper",
                },
                "ref": null,
                "rendered": "Paper",
                "type": "td",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Draw",
                },
                "ref": null,
                "rendered": "Draw",
                "type": "td",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Paper",
                },
                "ref": null,
                "rendered": "Paper",
                "type": "td",
              },
            ],
            "type": "tr",
          },
        ],
        "type": "table",
      },
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;

exports[`ScoreHistory no history renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ScoreHistory
    history={Array []}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <table
        className="gameHistory"
>
        
</table>,
      "className": "ScoreHistory",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [],
        "className": "gameHistory",
      },
      "ref": null,
      "rendered": Array [],
      "type": "table",
    },
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <table
          className="gameHistory"
>
          
</table>,
        "className": "ScoreHistory",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "className": "gameHistory",
        },
        "ref": null,
        "rendered": Array [],
        "type": "table",
      },
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
